### PR TITLE
Add support for new "vic-machine" flag to init Roles/Permission for ops-user, not exposed

### LINKED ISSUE #6505: VCH-Update: implementation of create/validate/repair of vSphere Permissions for ops-user and VCH resources
**User Statement:**
When a VCH is updated the ops-user should have Permissions on the various vSphere resources used by the VCH.
**Details:**
During VCH update the set of vSphere Permissions should be added for the appropriate Roles of the ops-user should be updated as well. The Permissions are applied to different vSphere managed resources that the VCH need to access and modify. The issue should track:

- Implementation of the code that updates the right Permission given the set of resources and the ops-user username,
- Unit tests,
- Manual basic integration testing with vCenter

**Acceptance Criteria:**
Code and tests should be implemented.

### PR DESCRIPTION
**Pull request by lcastellano**
In addition to the new flag, various bug fixes including the correct set of minimum permissions to run a container.

Fixes #6505 addendum

Issue: #6502 tracks the addition of the Nightly/Integration Test

### CODE REVIEW

File: cmd/vic-machine/common/ops_credentials.go [hunk 1]
```diff
@@ -47,6 +48,12 @@ func (o *OpsCredentials) Flags(hidden bool) []cli.Flag {
 			Usage:  "Password or token for the operations user. Defaults to the credential supplied with target",
 			Hidden: hidden,
 		},
+		cli.BoolFlag{
+			Name:        "ops-init-perms",
+			Usage:       "Initialize Roles and Permissions for ops-user",
```
Comment (anchal-agrawal) [inline cmd/vic-machine/common/ops_credentials.go]: `Initialize roles and permissions for ops-user` would be more consistent with the other usage texts above, what do you think?

File: lib/install/management/rbac.go [hunk 2]
```diff
@@ -281,7 +176,7 @@ func (am *AuthzManager) setupPermissions(ctx context.Context) ([]RBACResourcePer
 	}
 
 	// Find bridged networks
-	bridgeNet, ok := am.configSpec.Networks["bridge"]
+	bridgeNet, ok := am.configSpec.Network.ContainerNetworks["bridge"]
```
Comment (anchal-agrawal) [inline lib/install/management/rbac.go]: What's the purpose of this change?

Comment (lcastellano) [reply lib/install/management/rbac.go]: This is a bug, fix. Taking the bridgeNet from the Networks map works with the Sim but bot with real VC.

**Comment by lcastellano** [comment]
The integration test for this feature has to be done from scratch and requires creating users in VC which has not been done so far (no support in govmomi). This may require a substantial investigation/work. That's the reason why I broke it off in a different Issue (#6502) which is scheduled for the current sprint.

File: cmd/vic-machine/common/ops_credentials.go [hunk 3]
```diff
@@ -47,6 +48,12 @@ func (o *OpsCredentials) Flags(hidden bool) []cli.Flag {
 			Usage:  "Password or token for the operations user. Defaults to the credential supplied with target",
 			Hidden: hidden,
 		},
+		cli.BoolFlag{
+			Name:        "ops-init-perms",
```
Comment (anchal-agrawal) [inline cmd/vic-machine/common/ops_credentials.go]: Does `vic-machine configure` apply to this option? Ping @lcastellano @andrewtchin @mdubya66 - we made the call that for any new `vic-machine` option, `configure` support should be implemented simultaneously if applicable.

Comment (lcastellano) [reply cmd/vic-machine/common/ops_credentials.go]: Working with George on vic-machine-configure..

File: lib/config/virtual_container_host.go [hunk 4]
```diff
@@ -248,6 +253,14 @@ func (t *VirtualContainerHostConfigSpec) IsCreating() bool {
 	return t.ExecutorConfig.ID == CreatingVCH
 }
 
+func (t *VirtualContainerHostConfigSpec) SetGrantPerms() {
+	t.GrantPermsLevel = AddPerms
+}
+
+func (t *VirtualContainerHostConfigSpec) ShouldGrantPerms() bool {
+	return t.GrantPermsLevel != ""
```
Comment (anchal-agrawal) [inline lib/config/virtual_container_host.go]: Based on my understanding, this will always return true with this change, which would mean that `GrantOpsUserPerms` runs for every `configure` operation. Is that valid? I'd also like to see the `vic-machine configure` integration test passing with this change - please run specific ci.

Comment (lcastellano) [reply lib/config/virtual_container_host.go]: In the latest change I removed the flag from being available in vic-machine (both create and configure). I have done that for two reasons: need a better way to handle Boolean flags, lack of Integration tests. 

Eventually it will work this way:

Requesting the grant option in either create or configure, will do the granting and save the value of the option in the config for the VCH. This will be automatically be applied on each of the following configure request unless you explicitly specify --ops-grant-perms=false. Not specifying the flag will leave the situation as it was in the last create/configure operation. This is done to allow the user to omit the option in the following configure operations. Making it somewhat persistent. I have discussed with George at length about the approach and that is what we think is the best choice for now.

**Comment by lcastellano** [comment]
Since the integration test is not ready yet. I have removed the option from vic-machine, the rest of this PR is in preparation of the adding the new option + bug fixes + some refactoring.

**Comment by lcastellano** [comment]
Since the option is now removed from vic-machine this is definitely not the final merge.

File: lib/config/virtual_container_host.go (hunk 4 above)
Comment (anchal-agrawal) [reply lib/config/virtual_container_host.go]: > Not specifying the flag will leave the situation as it was in the last create/configure operation.

Could you point me to the code where this is handled? Thanks.

Also, if we're not enabling this feature in `vic-machine` yet, would it make sense to set `AddPerms` to `""` so `GrantOpsUserPerms` isn't invoked at all?

**Comment by mdubya66** [comment]
@mhagen-vmware with the vic-machine changes removed is this clear to merge with unit tests? @lcastellano you may want to change the name of the PR since it is no longer adding the vic-machine flag.

File: lib/config/virtual_container_host.go (hunk 4 above)
Comment (lcastellano) [reply lib/config/virtual_container_host.go]: The code that handles the different flavor of the Bool Flag is not in this commit. That is why I removed its external visibility. To support the three variants:
- the option is not set,
- the option is set,
- the option is set to false,
We need to implement a new version of the Bool flag, that's why I have removed it from now.
Regarding AddPerms I am not sure what you mean. GrantOpsUserPerms() is called only when GrantPermLevel is set to a non empty string. In the current version GrantPermLevel is always the empty string.

Comment (anchal-agrawal) [reply lib/config/virtual_container_host.go]: > In the current version GrantPermLevel is always the empty string.

We're saying `AddPerms = "ADD"` and `t.GrantPermsLevel = AddPerms`, so `ShouldGrantPerms()` will return true. Are we setting `AddPerms = ""` somewhere else?

**Comment by mhagen-vmware** [comment]
I don't see any unit tests?

**Comment by lcastellano** [comment]
I have added some unit tests. Some of the changes will only be testable when the option is exposed in vic-machine.

**Comment by anchal-agrawal** [comment]
Code LGTM. I'd suggest renaming the commit header to `Add support to init ops-user perms` - this indicates that while this change adds some functionality, it's not enabled in vic-machine (as there's no reference to vic-machine or its flags in the commit message)

File: lib/config/virtual_container_host.go (hunk 4 above)
Comment (hickeng) [reply lib/config/virtual_container_host.go]: I'm not sure if this is retreading the same ground as above. Was expecting this line to be:
```go
   return t.GrantPermsLevel == AddPerms
```

File: lib/install/management/configure.go [hunk 5]
```diff
@@ -111,6 +111,11 @@ func (d *Dispatcher) Configure(vch *vm.VirtualMachine, conf *config.VirtualConta
 	}
 
 	if err = d.update(conf, settings, isConfigureOp); err == nil {
+		if conf.ShouldGrantPerms() {
+			if err = GrantOpsUserPerms(d.ctx, d.session.Vim25(), conf); err != nil {
```
Comment (hickeng) [inline lib/install/management/configure.go]: Not going to block the PR for it, but I find the uncollapsed form of this is significantly easier to visually parse for important operations; the collapsed form needs extra effort to tell if there's a significant side effect or if it's just a test for existing state.
Only time I think it's really necessary to use the collapsed form is when scoping of the return is important:
```go
err := GrantOpsUserPerms(d.ctx, d.session.Vim25(), conf)
if err != nil {
	return errors.Errorf("Cannot init ops-user permissions, failure: %s. Exiting...", err)
}
```

File: lib/install/validate/validator_test_sim_util.go [hunk 6]
```diff
@@ -209,6 +209,9 @@ func (v *Validator) validateForSim(ctx context.Context, input *data.Data) (*conf
 	vmRef := vm.Reference()
 	conf.SetMoref(&vmRef)
 
+	// Remove bridge network from main section of config, to better emulate real VC behaviour
+	conf.Networks["bridge"] = nil
```
Comment (hickeng) [inline lib/install/validate/validator_test_sim_util.go]: I'm not sure why this needs doing - what was the bug that you mentioned that relates to this?

Comment (lcastellano) [reply lib/install/validate/validator_test_sim_util.go]: There are two changes related to this problem. The first is in rbac.go line 179. The original code was looking for the bridged network mo inside ExecutorConfig.Networks. That worked fine in the Simulator but failed when running against my VC test bed. For that to work I needed to change the rbac code to look at Network.ContainerNetworks. In case of the simulator the bridged network mo is found in both places. In case of my VC test bed it is only found in Network.ContainerNetworks. This modification tries to make the simulator behavior match the VC behavior, where ExecutorConfig.Networks can be nil.

**Comment by mdubya66** [comment]
@hickeng please review for config group
